text/0000-autoreborrow-traits.md

```diff
@@ -0,0 +1,468 @@
+  `ColRef`.
+- Kind: `MatMut` and `ColMut` are exclusive reference types. `MatRef` and
+  `ColRef` are shared reference types.
+- Usage: Enable various near-zero-cost operations over matrices by binding the
+  iteration to the original `Mat` through the reference types while statically
+  ensuring that indexed access is not required.
+
+Note: `Mat` is also reborrowable into `MatMut` and `MatRef`, but also has a
+`Drop` implementation. If reborrowing is understood to act on lifetimes of
+reference-like types, then reborrowing `Mat` is outside that definition as it
+neither has a lifetime nor is a reference-like type. It could be argued that
+`Mat` does have an implied `'static` lifetime, but the lack reference-like
+aspects of it at least cannot be denied. It might be that this use-case would be
+better understood as a generalised `AsRef` implementation.
+
+### [reborrow crate](https://docs.rs/reborrow/0.5.5/reborrow/index.html)
+
+This is a library used in eg. faer (above `Mat` and friends) that provides
+traits for emulating reborrowing.
+
+### Honorable mention: [CppRef and friends](https://docs.rs/autocxx/latest/autocxx/struct.CppRef.html)
+
+These types do not implement reborrowing, but cxx is interested in automatic
+coercion between the different C++ reference types. Since these are pointer
+wrapping types, some forms of autoreborrowing would make these coercions
+possible.
+
+## A note on `Reborrow` and `Copy`
+
+When dealing with shared references, reborrowing is not actually necessary. A
+shared reference ought to be `Copy` by virtue of being shareable, so for them
+implementing `Reborrow` is mostly unnecessary. The only reason that the
+`Reborrow` trait exists alongside `ReborrowMut` is because of the need to coerce
+exclusive references into shared references.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Reborrowing itself should be a fairly direct and "easy-to-grasp" feature in how
+it works; reference wrappers and custom reference types should be able to opt-in
+to reference-like reborrowing. This means that:
+
+1. When such a type is used as a function parameter, the type is implicitly
+   reborrowed.
+2. When such a type is used as a function parameter and a value bound to the
+   reborrowed type is returned, the lifetime is automatically expanded to the
+   calling context.
+3. When such a type `T` is returned from a function that returns a different
+   type `U` that `T` can be reborrowed into, the type is implicitly reborrowed.
+
+The implementation of reborrowing, however, is not an immediately obvious
+question. At least three main possibilities exist:
+
+## `Reborrow` and `ReborrowMut` traits with a GAT and methods
+
+Implement traits in the vein of the `reborrow` crate:
+
+```rust
+/// Immutable reborrowing.
+trait Reborrow<'short, _Outlives = &'short Self> {
+    type Target;
+
+    #[must_use]
+    fn rb(&'short self) -> Self::Target;
```

Comment:
Nit: Since these are never called explicitly, they might as well be spelled out as `reborrow`.